File: QuickSort/scala/Quicksort.scala
package com.jgdodson.algorithms

object Quicksort {

  def quicksort(arr: Array[Int]): Unit = {

    quicksort(arr, 0, arr.length-1)

  }

  def quicksort(arr: Array[Int], p: Int, r: Int): Unit = {

    if (p < r) {
      val q = partition(arr, p, r)
      quicksort(arr, p, q-1)
      quicksort(arr, q+1, r)
    }
  }

  // Partition the closed interval [p, r]
  private def partition(arr: Array[Int], p: Int, r: Int): Int = {

    var t = 0
    var i = p
    val pivot = arr(r);

    for (j <- p until r) {
      if (arr(j) < pivot) {

        // Swap
        t = arr(i)
        arr(i) = arr(j)
        arr(j) = t

        i += 1;
      }
    }

    // Swap pivot element into place
    t = arr(i)
    arr(i) = arr(r)
    arr(r) = t

    return i
  }

}

File: QuickSort/python/quicksort.py
'''
  AUTH: Jordan Dodson
  INFO: Implementation of quicksort in Python
'''


def quicksort(arr):
  _quicksort(arr, 0, len(arr)-1)


def _quicksort(arr, p, r):
  '''Sort the closed interval [p, r] recursively'''

  if p < r:
    q = partition(arr, p, r)
    _quicksort(arr, p, q-1)
    _quicksort(arr, q+1, r)


def partition(arr, p, r):

  pivot = arr[r]
  i = p

  while arr[i] < pivot:
    i += 1

  for j in range(i+1, r):
    if arr[j] < pivot:
      swap(arr, i, j)
      i += 1

  swap(arr, i, r)

  return i


def swap(arr, i, j):
  t = arr[i]
  arr[i] = arr[j]
  arr[j] = t
